pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,616 @@
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"assignPublicIP\" field, error = %v", err)
+	}
+
+	return &c, &pconfig, nil
+}
+
+func getVMIPAddresses(ctx context.Context, c *config, vm *compute.VirtualMachine) ([]string, error) {
+	var ipAddresses []string
+
+	if vm.VirtualMachineProperties == nil {
+		return nil, fmt.Errorf("machine is missing properties")
+	}
+
+	if vm.VirtualMachineProperties.NetworkProfile == nil {
+		return nil, fmt.Errorf("machine has no network profile")
+	}
+
+	if vm.NetworkProfile.NetworkInterfaces == nil {
+		return nil, fmt.Errorf("machine has no network interfaces data")
+	}
+
+	for n, iface := range *vm.NetworkProfile.NetworkInterfaces {
+		if iface.ID == nil || len(*iface.ID) == 0 {
+			return nil, fmt.Errorf("interface %d has no ID", n)
+		}
+
+		splitIfaceID := strings.Split(*iface.ID, "/")
+		ifaceName := splitIfaceID[len(splitIfaceID)-1]
+		addrs, err := getNICIPAddresses(ctx, c, ifaceName)
+		if vm.NetworkProfile.NetworkInterfaces == nil {
+			return nil, fmt.Errorf("failed to get addresses for interface %q: %s", ifaceName, err)
+		}
+		ipAddresses = append(ipAddresses, addrs...)
+	}
+
+	return ipAddresses, nil
+}
+
+func getNICIPAddresses(ctx context.Context, c *config, ifaceName string) ([]string, error) {
+	ifClient, err := getInterfacesClient(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create interfaces client: %v", err)
+	}
+
+	netIf, err := ifClient.Get(ctx, c.ResourceGroup, ifaceName, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get interface %q: %s", ifaceName, err.Error())
+	}
+
+	var ipAddresses []string
+
+	if netIf.IPConfigurations != nil {
+		for _, conf := range *netIf.IPConfigurations {
+			var name string
+			if conf.Name != nil {
+				name = *conf.Name
+			} else {
+				glog.Warning("IP configuration of NIC %q was returned with no name, trying to dissect the ID.", ifaceName)
+				if conf.ID == nil || len(*conf.ID) == 0 {
+					return nil, fmt.Errorf("IP configuration of NIC %q was returned with no ID", ifaceName)
+				}
+				splitConfID := strings.Split(*conf.ID, "/")
+				name = splitConfID[len(splitConfID)-1]
+			}
+
+			addrStrings, err := getIPAddressStrings(ctx, c, name)
+			if err != nil {
+				return nil, fmt.Errorf("failed to retrieve IP string for IP %q: %s", name, err)
+			}
+
+			ipAddresses = append(ipAddresses, addrStrings...)
+		}
+	}
+
+	return ipAddresses, nil
+}
+
+func getIPAddressStrings(ctx context.Context, c *config, addrName string) ([]string, error) {
+	ipClient, err := getIPClient(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create IP address client: %v", err)
+	}
+
+	ip, err := ipClient.Get(ctx, c.ResourceGroup, addrName, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IP %q: %s", addrName, err)
+	}
+
+	if ip.IPConfiguration == nil {
+		return nil, fmt.Errorf("IP %q has nil IPConfiguration", addrName)
```

Comment:
Please return lowercase errors only

Reply:
in case of `IP` I would be fine, but the linter will mark it also =/

Reply:
It doesn't mark it.

Reply:
But it will after #211

Reply:
It doesn't allow Errors starting with uppercase letters. We can keep it as-is, but will create more efforts later on.